WikiWatershed/model-my-watershed
# Custom Weather: Switch `custom_weather_data` to be FileField

## Overview

The initial assumption of `CharField` was incorrect, as the `django-storages` module being prepared in #3274 overrides Django's built-in File field.

As the migrations have not been applied yet on staging or other environments, it was alright to overwrite them.

Connects #3269 

### Demo

```http
❯ http --verbose --form :8000/mmw/modeling/projects/2/custom-weather-data/ weather@src/mmw/apps/modeling/templates/weather_data_sample.csv

POST /mmw/modeling/projects/2/custom-weather-data/ HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 680
Content-Type: multipart/form-data; boundary=7759d839747a56fff614862f235c4086
Host: localhost:8000
User-Agent: HTTPie/2.0.0

--7759d839747a56fff614862f235c4086
Content-Disposition: form-data; name="weather"; filename="weather_data_sample.csv"
Content-Type: text/csv

DATE,PRCP,TAVG
1/1/2001,0,22
1/2/2001,0,19
1/3/2001,0,18
1/4/2001,0,21
1/5/2001,0.04,25
1/6/2001,0.01,28
1/7/2001,0,33
1/8/2001,0.05,28
1/9/2001,0,19
1/10/2001,0,28
1/11/2001,0,36
1/12/2001,0,30
1/13/2001,0,29
1/14/2001,0,29
1/15/2001,0,34
1/16/2001,0,34
1/17/2001,0,31
1/18/2001,0.01,30
1/19/2001,0.4,31
1/20/2001,0.21,27
1/21/2001,0,18
1/22/2001,0,17
1/23/2001,0,19
1/24/2001,0,32
1/25/2001,0,22
1/26/2001,0,20
1/27/2001,0,30
1/28/2001,0,23
1/29/2001,0,23
1/30/2001,0.57,38
1/31/2001,0.01,37

--7759d839747a56fff614862f235c4086--

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Mon, 30 Mar 2020 21:09:05 GMT
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie

{
    "area_of_interest_name": "Lower Schuylkill River, HUC-10 Watershed (ID 0204020310)",
    "created_at": "2019-08-28T13:57:19.915671-04:00",
    "hydroshare": null,
    "id": 2,
    "is_private": true,
    "model_package": "gwlfe",
    "modified_at": "2020-03-30T17:09:05.933784-04:00",
    "name": "Subbasin Legend",
    "user": 2,
    "uses_custom_weather": true
}
```

### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

* Check out this branch, and `manage dbshell`
* Run these queries to undo the last migration:

    ```sql
    ALTER TABLE modeling_project
        DROP COLUMN custom_weather_dataset,
        DROP COLUMN uses_custom_weather;

    DELETE FROM django_migrations
        WHERE name = '0030_custom_weather_dataset';
    ```

* Apply the new migration `manage migrate`
* Apply this patch: [test-3276.patch.txt](https://github.com/WikiWatershed/model-my-watershed/files/4405527/test-3276.patch.txt)
* Ensure you have at least one saved project in your local database
* Upload the sample file to the project, replacing `2` with your project's ID:

    ```http
    http --verbose --form :8000/mmw/modeling/projects/2/custom-weather-data/ weather@src/mmw/apps/modeling/templates/weather_data_sample.csv
    ```

    - [x] Ensure you see a response similar to the Demo above, specifically with `uses_custom_weather` set to `true`
    - [ ] Ensure you see the file inside the VM at `/var/www/mmw/media/project_2/`, again replacing `2` with your chosen project's ID.
